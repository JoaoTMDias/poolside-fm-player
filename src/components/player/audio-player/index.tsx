/* eslint-disable jsx-a11y/media-has-caption */
// Libraries
import * as React from "react";
import ControlsMedia from "../controls-media/index";
import { IAudioPlayerProps, IAudioPlayerState, EPlayingStatus } from "../player.interfaces";

/**
 * @description Component Description
 * @author Jo√£o Dias
 * @date 2019-02-16
 * @returns {React.FunctionComponent<IAudioPlayerProps>}
 */
class AudioPlayer extends React.Component<IAudioPlayerProps, IAudioPlayerState> {
	private player: React.RefObject<HTMLAudioElement>;

	constructor(props: IAudioPlayerProps) {
		super(props);

		this.state = {
			player: null,
			status: EPlayingStatus.paused,
		};

		// Binds
		this.onClickOnPrevious = this.onClickOnPrevious.bind(this);
		this.onClickOnPlay = this.onClickOnPlay.bind(this);
		this.onClickOnNext = this.onClickOnNext.bind(this);

		// Refs
		this.player = React.createRef<HTMLAudioElement>();
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	componentDidMount() {
		const { player } = this.state;
		if (this.player.current) {
			this.setState(
				{
					player: this.player.current,
				},
				() => {
					if (player) {
						player.currentTime = 0;
						player.pause();
					}
				},
			);
		}
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	onClickOnPrevious() {
		console.log("previous");
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	onClickOnNext() {
		console.log("next");
	}

	/**
	 *
	 *
	 * @memberof AudioPlayer
	 */
	onClickOnPlay() {
		const { status } = this.state;

		let nextStatus;

		switch (status) {
			default:
			case EPlayingStatus.paused:
				nextStatus = EPlayingStatus.playing;
				break;

			case EPlayingStatus.playing:
				nextStatus = EPlayingStatus.paused;
		}

		this.handleAudioPlayer(nextStatus);
	}

	/**
	 *
	 *
	 * @param {EPlayingStatus} status
	 * @memberof AudioPlayer
	 */
	handleAudioPlayer(status: EPlayingStatus) {
		const { player } = this.state;

		if (player) {
			switch (status) {
				case EPlayingStatus.playing:
					player.play();
					break;

				case EPlayingStatus.paused:
					player.pause();
					break;

				default:
					break;
			}

			this.updatePlayingStatus(status);
		}
	}

	/**
	 *
	 *
	 * @param {EPlayingStatus} status
	 * @memberof AudioPlayer
	 */
	updatePlayingStatus(status: EPlayingStatus) {
		this.setState({
			status,
		});
	}

	render() {
		const { status } = this.state;
		const { src } = this.props;

		return (
			<div id="media-player" className="media-player row">
				<audio
					ref={this.player}
					id="audio-player"
					src={src}
					data-testid="component-audio-player"
					autoPlay={false}
					className="sr-only"
				/>
				<ControlsMedia
					status={status}
					onClickOnPrevious={this.onClickOnPrevious}
					onClickOnPlay={this.onClickOnPlay}
					onClickOnNext={this.onClickOnNext}
				/>
			</div>
		);
	}
}

export default React.memo(AudioPlayer);
