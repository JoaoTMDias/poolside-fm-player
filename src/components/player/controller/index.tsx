import React, { useRef, useState, useReducer } from "react";
import SoundCloudAudio from "soundcloud-audio";
import { PoolsidePlaylists } from "data/constants";
import {
	PlayerControllerContext,
	defaultPlayerControllerState,
	IPlayerControllerContext,
} from "contexts/player-controller-context";
import { useDidMount } from "helpers/custom-hooks/use-lifecycle-hooks";
import {
	ISoundcloudPlayer,
	EPlayingStatus,
	ISoundcloudPlaylist,
	ESoundCloudPlayerEvents,
	IMediaPlayerTrackMetadata,
} from "../media-player/player.interfaces";
import { updateTrackReducer } from "./update-tracker-reducer";
import { updateStatusReducer } from "./update-status-reducer";
import { isNil } from "helpers";

/**
 * Get a random number between a range
 *
 * @param {number} allTracks
 * @returns
 */
export function getRandomTrackIndex(allTracks: number) {
	const min = Math.ceil(0);
	const max = Math.floor(allTracks);
	return Math.floor(Math.random() * (max - min)) + min;
}

export interface IPlayerControllerProps {
	onPrevious?: (hasChanged: boolean) => void;
	onNext?: (hasChanged: boolean) => void;
	onPlay?: (status: EPlayingStatus) => void;
	onChangeVolume?: (volume: number) => void;
	children: React.ReactNode;
}

/**
 * PlayerController
 *
 * Controls the player instance.
 * Plays, pauses, changes tracks and then updates the state and sends it through the Context API.
 *
 */
const PlayerController: React.FunctionComponent<IPlayerControllerProps> = ({
	onPrevious,
	onNext,
	onPlay,
	onChangeVolume,
	children,
}) => {
	const { current: player } = useRef<ISoundcloudPlayer>(
		new SoundCloudAudio(process.env.REACT_APP_API_KEY)
	);
	const [artist, setArtist] = useState(defaultPlayerControllerState.artist);
	const [title, setTitle] = useState(defaultPlayerControllerState.title);
	const [duration] = useState(defaultPlayerControllerState.duration);
	const [currentTime] = useState(defaultPlayerControllerState.currentTime);
	const [currentIndex, setCurrentIndex] = useState(defaultPlayerControllerState.currentIndex);
	const [track, updateTrack] = useReducer(updateTrackReducer, defaultPlayerControllerState.track);
	const [status, updateStatus] = useReducer(
		updateStatusReducer,
		defaultPlayerControllerState.status
	);

	/**
	 * Handles the click on the previous button
	 *
	 * @returns {boolean}
	 */
	function _onPrevious(): boolean {
		let hasChanged = false;

		if (player && track) {
			const { current, last } = track;
			const previousIndex = current <= 0 ? last : current - 1;

			updateTrack({
				type: "CURRENT",
				payload: previousIndex,
			});

			if (player?.playing && player?.previous) {
				player.previous();
			}

			hasChanged = true;
		}

		if (onPrevious) {
			onPrevious(hasChanged);
		}

		return hasChanged;
	}

	/**
	 * Handles the click on the next button
	 *
	 * @returns {boolean}
	 */
	function _onNext(): boolean {
		let hasChanged = false;

		if (player && track) {
			const { current, last } = track;

			const nextIndex = current === last ? 0 : current + 1;

			updateTrack({
				type: "CURRENT",
				payload: nextIndex,
			});

			if (player?.next && player?.playing) {
				player.next();
			}

			hasChanged = true;
		}

		if (onNext) {
			onNext(hasChanged);
		}

		return hasChanged;
	}

	/**
	 * Handles the new status.
	 *
	 * @returns {void}
	 */
	function handleNewStatus(): void {
		if (player && player.play && player.pause) {
			switch (status) {
				case EPlayingStatus.playing:
					player.pause();
					break;

				case EPlayingStatus.ready:
				case EPlayingStatus.paused:
					player.play({
						playlistIndex: track.current,
					});
					break;

				default:
					break;
			}
		}
	}

	/**
	 * Callback for handling the clicking on the play button
	 *
	 * @returns {EPlayingStatus | null}
	 */
	function _onPlay(): void {
		switch (status) {
			case EPlayingStatus.ready:
			case EPlayingStatus.paused:
				updateStatus({
					type: "PLAY",
				});
				break;

			case EPlayingStatus.playing:
				updateStatus({
					type: "PAUSE",
				});
				break;

			default:
			case EPlayingStatus.idle:
			case EPlayingStatus.loading:
			case EPlayingStatus.error:
				break;
		}

		handleNewStatus();

		if (onPlay) {
			onPlay(status);
		}
	}

	/**
	 * Changes the volume of the current playing audio
	 *
	 * @param {number} volume
	 */
	function _onChangeVolume(volume: number) {
		const { audio } = player;
		if (player && audio && volume) {
			audio.volume = volume;

			if (onChangeVolume) {
				onChangeVolume(volume);
			}
		}
	}

	/**
	 * Returns the current tracks title and artist
	 *
	 * @param {number} index
	 * @returns {(IMediaPlayerTrackMetadata | null)}
	 */
	function getCurrentTrackAndArtist(index: number): IMediaPlayerTrackMetadata | null {
		if (index && player?._playlist) {
			// eslint-disable no-underscore-dangle
			const { title: currentTitle } = player._playlist.tracks[index];
			const { user } = player._playlist.tracks[index];
			// eslint-enable no-underscore-dangle

			const currentArtist = user && user.username;

			if (title && artist) {
				return {
					title: currentTitle,
					artist: currentArtist,
				};
			}
		}

		return null;
	}

	/**
	 * Loads a new playlist
	 *
	 * @param {ISoundcloudPlaylist} playlist
	 */
	async function loadPlaylist(playlist: ISoundcloudPlaylist) {
		const current = getRandomTrackIndex(playlist.track_count);
		const last = playlist.track_count - 1;

		updateStatus({
			type: "LOAD",
		});

		if (player && player.on && player) {
			// once playlist is loaded it can be played
			updateTrack({
				type: "TRACK",
				payload: {
					current,
					last,
				},
			});

			updateStatus({
				type: "READY",
			});

			player.on(ESoundCloudPlayerEvents.canplay, () => {
				const { _playlistIndex } = player;
				const metadata = _playlistIndex ? getCurrentTrackAndArtist(_playlistIndex) : null;

				if (metadata) {
					setTitle(metadata.title);
					setArtist(metadata.artist);
				}
			});

			// for playlists it's possible to switch to another track in queue
			// e.g. we do it here when playing track is finished
			player.on(ESoundCloudPlayerEvents.ended, () => {
				_onNext();
			});
		}
	}

	/**
	 * get the current playlist and starts a new playlist instance
	 *
	 * @returns {void}
	 */
	function startPlaylist(): void {
		const currentPlaylist = PoolsidePlaylists[currentIndex].url;

		if (player?.resolve && currentPlaylist) {
			player.resolve(currentPlaylist, (playlist: ISoundcloudPlaylist) => {
				loadPlaylist(playlist);
			});
		}
	}

	/**
	 * When the user selects a new playlist, updates the current playlist index state
	 *
	 * @param {number} index
	 * @returns {void}
	 */
	function _onChangeOption(index: number): void {
		if (!isNil(index) && player && player.stop) {
			player.stop();

			setCurrentIndex(index);
			startPlaylist();
		}
	}

	useDidMount(() => {
		if (player) {
			startPlaylist();
		}
	});

	const contextValue: IPlayerControllerContext = {
		status,
		artist,
		title,
		duration,
		currentTime,
		currentIndex,
		track,
		audio: player.audio,
		previous: () => _onPrevious(),
		next: () => _onNext(),
		togglePlay: () => _onPlay(),
		changeVolume: (volume) => _onChangeVolume(volume),
		onChangeOption: (index) => _onChangeOption(index),
	};

	return (
		<PlayerControllerContext.Provider value={contextValue}>
			{children}
		</PlayerControllerContext.Provider>
	);
};

export default PlayerController;
