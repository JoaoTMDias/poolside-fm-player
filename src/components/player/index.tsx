import ControlsMedia from "./controls-media/index";
import AudioPlayer from "./audio-player/index";

export { AudioPlayer, ControlsMedia };
