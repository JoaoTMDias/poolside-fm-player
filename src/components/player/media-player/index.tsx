/* eslint-disable jsx-a11y/media-has-caption */
// Libraries
import * as React from "react";
import { PlayerControllerContext } from "contexts/player-controller-context";
import CurrentSong from "../../current-song/index.component";
import ControlsMedia from "../controls-media/index";

/**
 * @description Component Description
 * @author Jo√£o Dias
 * @date 2019-02-16
 * @returns {React.FunctionComponent<IMediaPlayerProps>}
 */
const MediaPlayer = () => {
	return (
		<PlayerControllerContext.Consumer>
			{({
				artist,
				changeVolume,
				currentTime,
				next,
				previous,
				status,
				title,
				togglePlay,
				track,
			}) => {
				return (
					<div
						id="media-player"
						data-testid="component-media-player"
						className="media-player row"
						data-current-track={track?.current.toString()}
						data-playing-status={status}
					>
						<CurrentSong title={title} artist={artist} currentTime={`${currentTime}`} />
						<ControlsMedia
							status={status}
							onClickOnPrevious={previous}
							onTogglePlay={togglePlay}
							onClickOnNext={next}
							onChangeVolume={changeVolume}
						/>
					</div>
				);
			}}
		</PlayerControllerContext.Consumer>
	);
};

export default MediaPlayer;
