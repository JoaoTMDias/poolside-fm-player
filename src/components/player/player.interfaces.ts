export enum EPlayingStatus {
	playing = "playing",
	paused = "paused",
}

export enum ESoundCloudPlayerEvents {
	ended = "ended",
	canplay = "canplay",
	load = "load",
	loadeddata = "loadeddata",
	loadedmetadata = "loadedmetadata",
	loadend = "loadend",
	loadstart = "loadstart",
	pause = "pause",
	play = "play",
	playing = "playing",
	seeked = "seeked",
	seeking = "seeking",
	timeupdate = "timeupdate",
	volumechange = "volumechange",
	waiting = "waiting",
}

export enum ESoundcloudPlayerPreloadType {
	none = "none",
	metadata = "metadata",
	auto = "auto",
}

export interface IMediaPlayerProps {
	currentPlaylistIndex: number;
}

export interface IMediaPlayerState {
	ready: boolean;
	status: EPlayingStatus;
}

export interface IControlsMediaProps {
	onClickOnPrevious?: () => void;
	onTogglePlay?: () => void;
	onClickOnNext?: () => void;
	onClickOnVolume?: () => void;
	status: EPlayingStatus;
}

interface ISoucloudPlayerMethodsPlay {
	streamUrl?: string;
	playlistIndex?: number;
}

export interface ISoundcloudPlayer {
	events: any;
	_baseUrl: string;
	_clientId: string;
	_playlist: ISoundcloudPlaylist;
	_playlistIndex: 0;
	_track: any;

	/**
	 * Instance of raw <audio> element. There are several useful properties like currentTime (in seconds) or events
	 * you may want to listen with addEventListener (the full list of of them at
	 * https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement).
	 */
	audio: HTMLAudioElement;

	/**
	 * SoundCloud track duration converted into seconds in order to be in sync with audio.currentTime.
	 */
	duration: number;

	/**
	 * Shows the current state of the player, returns false or source of a currently streaming track.

	 */
	playing: string;

	/**
	 * If you don't have SoundCloud stream_url (e.g. https://api.soundcloud.com/tracks/:id/stream) or
	 * you need track's metadata then this method is for you. Pass original track's or playlist's url as a first argument.
	 * Once data will be resolved without errors, callback function will receive it as plain object as the only argument.
	 */
	resolve: (url: string, callback: (playlist: ISoundcloudPlaylist) => void) => void;

	/**
	 * Start playing track if it's not playing right now.
	 * Returns a Promise and accepts options object:
	 * - options.streamUrl - any audio streaming url string (e.g. SoundCloud track's stream_url),
	 * if it's passed it will be the main playing source.
	 * - options.playlistIndex - number that specifies the position of the track to play in resolved
	 * SoundCloud playlist's tracks array.
	 */
	play: (options?: ISoucloudPlayerMethodsPlay) => Promise<any>;

	/**
	 * Pause playing audio.
	 */
	pause: () => void;

	/**
	 * Stop playing audio and rewind it to start.
	 */
	stop: () => void;

	/**
	 * Skip to the next track in playlist to play.
	 * Returns a Promise and accepts options object:
	 * - options.loop - boolean, if set to true will start at the beginning of a playlist after the last track.
	 */
	next: (options?: { loop: boolean }) => Promise<any>;

	/**
	 * Return to the previous track in playlist (returns a Promise).
	 */
	previous: () => Promise<any>;

	/**
	 * Preload track data without playing it.
	 *
	 * - preloadType - this attribute is intended to provide a hint to the browser about what the author thinks will lead
	 * to the best user experience. It may have one of the following values:
	 * 	1. 'none' - indicates that the audio should not be preloaded
	 * 	2. 'metadata' - indicates that only audio metadata (e.g. length) is fetched
	 * 	3. 'auto' - indicates that the whole audio file could be downloaded, even if the user is not expected to use it
	 *
	 * see more at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio#Attributes
	 */
	preload: (streamUrl: string, preloadType: ESoundcloudPlayerPreloadType) => void;
	seek: any;
	on: (event: ESoundCloudPlayerEvents, handler: () => void) => void;
	off: (event: ESoundCloudPlayerEvents, handler: () => void) => void;
	unbindAll: () => void;
}

export interface ISoundcloudPlaylist {
	duration: number;
	release_day: null;
	permalink_url: string;
	reposts_count: number;
	genre: null;
	permalink: string;
	purchase_url: null;
	release_month: null;
	description: null;
	uri: string;
	label_name: null;
	tag_list: string;
	release_year: null;
	track_count: number;
	user_id: number;
	last_modified: string;
	ISoundcloudLicense: ISoundcloudLicense;
	tracks: ISoundcloudTrack[];
	playlist_type: null;
	id: number;
	downloadable: null;
	sharing: ISoundcloudSharing;
	created_at: string;
	release: null;
	likes_count: number;
	kind: string;
	title: string;
	type: null;
	purchase_title: null;
	artwork_url: null;
	ean: null;
	streamable: boolean;
	user: ISoundcloudUser;
	embeddable_by: ISoundcloudEmbeddableBy;
	label_id: null;
}

export enum ISoundcloudEmbeddableBy {
	All = "all",
}

export enum ISoundcloudLicense {
	AllRightsReserved = "all-rights-reserved",
	CcBy = "cc-by",
	CcByNc = "cc-by-nc",
	CcByNcSa = "cc-by-nc-sa",
}

export enum ISoundcloudSharing {
	Public = "public",
}

export interface ISoundcloudTrack {
	comment_count: number;
	downloadable: boolean;
	release: null | string;
	created_at: string;
	description: null | string;
	original_content_size: number;
	title: string;
	track_type: null | string;
	duration: number;
	video_url: null;
	original_format: ISoundcloudOriginalFormat;
	artwork_url: null | string;
	streamable: boolean;
	tag_list: string;
	release_month: number | null;
	genre: string;
	release_day: number | null;
	id: number;
	state: ISoundcloudState;
	reposts_count: number;
	last_modified: string;
	label_name: null | string;
	commentable: boolean;
	bpm: null;
	policy: ISoundcloudPolicy;
	favoritings_count: number;
	kind: TrackKind;
	purchase_url: null | string;
	release_year: number | null;
	key_signature: null | string;
	isrc: null | string;
	sharing: ISoundcloudSharing;
	uri: string;
	attachments_uri: string;
	download_count: number;
	likes_count: number;
	license: ISoundcloudLicense;
	purchase_title: null | string;
	user_id: number;
	embeddable_by: ISoundcloudEmbeddableBy;
	monetization_model: ISoundcloudMonetizationModel;
	waveform_url: string;
	permalink: string;
	permalink_url: string;
	user: ISoundcloudUser;
	label_id: null;
	stream_url: string;
	playback_count: number;
	download_url?: string;
}

export enum TrackKind {
	Track = "track",
}

export enum ISoundcloudMonetizationModel {
	AdSupported = "AD_SUPPORTED",
	Blackbox = "BLACKBOX",
	NotApplicable = "NOT_APPLICABLE",
}

export enum ISoundcloudOriginalFormat {
	Aiff = "aiff",
	Mp3 = "mp3",
	Wav = "wav",
}

export enum ISoundcloudPolicy {
	Allow = "ALLOW",
	Monetize = "MONETIZE",
}

export enum ISoundcloudState {
	Finished = "finished",
}

export interface ISoundcloudUser {
	id: number;
	kind: ISoundcloudUserKind;
	permalink: string;
	username: string;
	last_modified: string;
	uri: string;
	permalink_url: string;
	avatar_url: string;
}

export enum ISoundcloudUserKind {
	User = "user",
}
