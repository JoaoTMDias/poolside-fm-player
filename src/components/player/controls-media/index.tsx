// Libraries
import * as React from "react";
import { Link } from "react-router-dom";
import { ROUTE_SETTINGS } from "data/constants/routes";
import { hasPressedSpaceOrEnter } from "helpers";
import * as S from "./controls-media.styled";
import { IControlsMediaProps, EPlayingStatus } from "../media-player/player.interfaces";
import ButtonVolume from "./buttons/button-volume";

/**
 * @description Controls - Media Buttons
 * @author Jo√£o Dias
 * @date 2019-02-16
 * @returns {React.FunctionComponent<IControlsMediaProps>}
 */
const ControlsMedia: React.FunctionComponent<IControlsMediaProps> = ({
	onClickOnPrevious,
	onTogglePlay,
	onClickOnNext,
	onClickOnVolume,
	status,
}) => {
	return (
		<S.Wrapper id="controls" className="controls">
			<S.Buttons id="controls-media" className="controls__media">
				<li className="controls-media__item controls-media__item--first">
					<button
						id="controls-media-button-previous"
						data-testid="component-controls-media-button-previous"
						title="Previous Song"
						aria-label="Click play the previous song"
						type="button"
						className="controls-media__button button button--large"
						onClick={() => {
							if (onClickOnPrevious) {
								onClickOnPrevious();
							}
						}}
						onKeyUp={(event: React.KeyboardEvent<HTMLButtonElement>) => {
							if (onClickOnPrevious && hasPressedSpaceOrEnter(event.keyCode)) {
								onClickOnPrevious();
							}
						}}
					>
						<svg
							className="icon"
							xmlns="http://www.w3.org/2000/svg"
							width="12"
							height="9"
							fill="none"
							viewBox="0 0 12 9"
						>
							<path
								fill="var(--color-icon, #000)"
								d="M12 0v9h-1V8h-1V7H9V6H8V5H7v4H6V8H5V7H4V6H3V5H2v4H0V0h2v4h1V3h1V2h1V1h1V0h1v4h1V3h1V2h1V1h1V0h1z"
							/>
						</svg>
					</button>
				</li>
				<li className="controls-media__item controls-media__item--middle">
					<button
						id="controls-media-button-play"
						data-testid="component-controls-media-button-play"
						title="Play song"
						aria-label="Click play the song"
						type="button"
						className="controls-media__button button button--large"
						onClick={() => {
							if (onTogglePlay) {
								onTogglePlay();
							}
						}}
						onKeyUp={(event: React.KeyboardEvent<HTMLButtonElement>) => {
							if (onTogglePlay && hasPressedSpaceOrEnter(event.keyCode)) {
								onTogglePlay();
							}
						}}
					>
						{status === EPlayingStatus.paused ? (
							<svg
								id="icon-pause"
								className="icon"
								xmlns="http://www.w3.org/2000/svg"
								width="9"
								height="9"
								fill="none"
								viewBox="0 0 9 9"
							>
								<path fill="var(--color-icon, #000)" d="M3 9V0h1v1h1v1h1v1h1v1h1v1H7v1H6v1H5v1H4v1H3z" />
							</svg>
						) : (
							<svg
								id="icon-play"
								className="icon"
								width="9"
								height="9"
								viewBox="0 0 9 9"
								fill="none"
								xmlns="http://www.w3.org/2000/svg"
							>
								<path d="M2 0H4V9H2V0Z" fill="var(--color-icon, #000)" />
								<path d="M5 0H7V9H5V0Z" fill="var(--color-icon, #000)" />
							</svg>
						)}
					</button>
				</li>
				<li className="controls-media__item controls-media__item--last">
					<button
						id="controls-media-button-next"
						data-testid="component-controls-media-button-next"
						title="Next Song"
						aria-label="Click play the next song"
						type="button"
						className="controls-media__button button button--large"
						onClick={() => {
							if (onClickOnNext) {
								onClickOnNext();
							}
						}}
						onKeyUp={(event: React.KeyboardEvent<HTMLButtonElement>) => {
							if (onClickOnNext && hasPressedSpaceOrEnter(event.keyCode)) {
								onClickOnNext();
							}
						}}
					>
						<svg
							className="icon"
							xmlns="http://www.w3.org/2000/svg"
							width="12"
							height="9"
							fill="none"
							viewBox="0 0 12 9"
						>
							<path
								fill="var(--color-icon, #000)"
								d="M0 9V0h1v1h1v1h1v1h1v1h1V0h1v1h1v1h1v1h1v1h1V0h2v9h-2V5H9v1H8v1H7v1H6v1H5V5H4v1H3v1H2v1H1v1H0z"
							/>
						</svg>
					</button>
				</li>
			</S.Buttons>
			<div className="controls-media__item">
				<ButtonVolume onClickOnVolume={onClickOnVolume} />
			</div>
			<div className="controls-media__item">
				<Link
					to={ROUTE_SETTINGS}
					id="controls-button-settings"
					data-testid="component-controls-button-settings"
					title="Settings"
					aria-label="Click to navigate to the app's settings page"
					className="controls-media__button button button--large"
				>
					<svg className="icon" xmlns="http://www.w3.org/2000/svg" width="9" height="9" fill="none" viewBox="0 0 9 9">
						<path
							fill="var(--color-icon, #000)"
							d="M8.413 3.913v-.587h-.587V2.152h.587v-.978h-.587V.587h-.978v.587H5.674V.587h-.587V0H3.913v.587h-.587v.587H2.152V.587h-.978v.587H.587v.978h.587v1.174H.587v.587H0v1.174h.587v.587h.587v1.174H.587v.978h.587v.587h.978v-.587h1.174v.587h.587V9h1.174v-.587h.587v-.587h1.174v.587h.978v-.587h.587v-.978h-.587V5.674h.587v-.587H9V3.913h-.587zM6.065 5.478h-.587v.587H3.522v-.587h-.587V3.522h.587v-.587h1.956v.587h.587v1.956z"
						/>
					</svg>
				</Link>
			</div>
		</S.Wrapper>
	);
};

ControlsMedia.defaultProps = {
	status: EPlayingStatus.paused,
};

export default React.memo(ControlsMedia);
