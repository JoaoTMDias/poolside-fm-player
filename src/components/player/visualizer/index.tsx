// Libraries
import * as React from "react";
import { PlayerControllerContext, IPlayerControllerState } from "contexts/player-controller-context";
import { EPlayingStatus } from "components/player/media-player/player.interfaces";
import Preview from "assets/images/now-playing.gif";
import * as S from "./player-visualizer.styled";

/**
 * @description An audio visualizer for the player
 * @author JoÃ£o Dias
 * @date 2019-02-16
 * @returns {React.FunctionComponent}
 */
const PlayerVisualizer: React.FunctionComponent = () => {
	return (
		<PlayerControllerContext.Consumer>
			{(state: IPlayerControllerState) => (
				<S.PlayerVisualizerWrapper
					role="presentation"
					data-testid="player-visualizer-wrapper"
					id="player-visualizer"
					className={`player-visualizer ${state.status === EPlayingStatus.playing && "player-visualizer--is-playing"}`}
				>
					<figcaption className="sr-only">This is an audio visualizer for the music that is playing</figcaption>
					<img className="player-visualizer__image" src={Preview} width="272" height="32" alt="Audio visualizer" />
				</S.PlayerVisualizerWrapper>
			)}
		</PlayerControllerContext.Consumer>
	);
};

export default React.memo(PlayerVisualizer);
