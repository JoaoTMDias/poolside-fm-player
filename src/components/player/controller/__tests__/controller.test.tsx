import React from "react";
import { shallow, mount, ReactWrapper } from "enzyme";
import PlayerController, { getRandomTrackIndex } from "components/player/controller/index";
import PlayerVisualizer from "components/player/visualizer";
import { IPlayerControllerState } from "contexts/player-controller-context";
import { EPlayingStatus } from "components/player/media-player/player.interfaces";

// Audio Prototype hacks
window.HTMLMediaElement.prototype.pause = () => {
	/* do nothing */
};

describe("<PlayerController />", () => {
	describe("render", () => {
		it("should render the controller", () => {
			const component = shallow(<PlayerController />);

			expect(component).toMatchSnapshot();
		});
	});

	describe("getRandomTrackIndex", () => {
		it("should return a random number", () => {
			const randomNumber = getRandomTrackIndex(200);
			expect(typeof randomNumber).toBe("number");
		});
	});

	describe("onClickOnPrevious", () => {
		it("should change the current track number", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);

			const PlayerControllerInstance = wrapper.instance();
			const track = {
				current: 2,
				last: 200,
			};

			wrapper.setState({
				track,
			});

			PlayerControllerInstance.onClickOnPrevious(track);

			expect(wrapper.state().track.current).toBe(1);
		});

		it("should change the current track number to the last one", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);

			const PlayerControllerInstance = wrapper.instance();

			const track = {
				current: 0,
				last: 200,
			};

			wrapper.setState({
				track,
			});

			PlayerControllerInstance.onClickOnPrevious(track);

			expect(wrapper.state().track.current).toBe(200);
		});
	});

	describe("onClickOnNext", () => {
		it("should change the current track number", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);

			const PlayerControllerInstance = wrapper.instance();
			const track = {
				current: 2,
				last: 50,
			};

			wrapper.setState({
				track,
			});

			PlayerControllerInstance.onClickOnNext(track);

			expect(wrapper.state().track.current).toEqual(3);
		});

		it("should change the current track number to the first item in the array index", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);

			const instance = wrapper.instance();
			const track = {
				current: 200,
				last: 200,
			};

			wrapper.setState({
				track,
			});

			instance.onClickOnNext(track);

			expect(wrapper.state().track.current).toEqual(0);
		});
	});

	describe("onTogglePlay", () => {
		it("should switch to play by default", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);
			const mock = jest.fn();

			const instance = wrapper.instance();
			instance.handleMediaPlayer = mock;

			instance.onTogglePlay();

			expect(mock).toHaveBeenCalledWith(EPlayingStatus.playing);
		});

		it("should switch to play if paused", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);
			const mock = jest.fn();

			const instance = wrapper.instance();
			instance.handleMediaPlayer = mock;

			instance.onTogglePlay(EPlayingStatus.paused);

			expect(mock).toHaveBeenCalledWith(EPlayingStatus.playing);
		});

		it("should switch to paused if playing", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);
			const mock = jest.fn();

			const instance = wrapper.instance();
			instance.handleMediaPlayer = mock;

			instance.onTogglePlay(EPlayingStatus.playing);

			expect(mock).toHaveBeenCalledWith(EPlayingStatus.paused);
		});
	});

	describe("onChangePlaylist", () => {
		it("should stop playing when called", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount<PlayerController>(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);

			const instance = wrapper.instance();

			instance.onChangePlaylist(1);

			expect(wrapper.state().currentPlaylistIndex).toBe(1);
			expect(wrapper.instance().player.playing).toBe(false);
		});
	});

	describe("updatePlayingStatus", () => {
		it("should update the state with the new playing status", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount<PlayerController>(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);

			const instance = wrapper.instance();

			instance.updatePlayingStatus(EPlayingStatus.playing);

			expect(wrapper.state().status).toBe(EPlayingStatus.playing);
		});
	});

	describe("handleMediaPlayer", () => {
		it("should call updatePlayingStatus", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount<PlayerController>(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);
			const mock = jest.fn();
			const instance = wrapper.instance();

			instance.updatePlayingStatus = mock;

			instance.handleMediaPlayer(EPlayingStatus.paused);

			expect(mock).toHaveBeenCalledWith(EPlayingStatus.paused);
		});

		it("should not call play or pause", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount<PlayerController>(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);
			const mockPlay = jest.fn();
			const mockPause = jest.fn();
			const instance = wrapper.instance();

			instance.player.play = mockPlay;
			instance.player.pause = mockPause;

			instance.handleMediaPlayer();

			expect(mockPlay).not.toHaveBeenCalled();
			expect(mockPause).not.toHaveBeenCalled();
		});

		it("should play a song", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount<PlayerController>(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);
			const mock = jest.fn();
			const instance = wrapper.instance();

			instance.player.play = mock;

			instance.handleMediaPlayer(EPlayingStatus.playing);

			expect(mock).toHaveBeenCalled();
		});

		it("should paused a song", () => {
			const wrapper: ReactWrapper<{}, IPlayerControllerState> = mount<PlayerController>(
				<PlayerController>
					<PlayerVisualizer />
				</PlayerController>,
			);
			const mock = jest.fn();
			const instance = wrapper.instance();

			instance.player.pause = mock;

			instance.handleMediaPlayer(EPlayingStatus.paused);

			expect(mock).toHaveBeenCalled();
		});
	});
});
