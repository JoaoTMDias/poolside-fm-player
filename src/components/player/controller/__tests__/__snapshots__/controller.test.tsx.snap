// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerController /> render should render the controller 1`] = `ShallowWrapper {}`;
